### PR TITLE
Pufferfish/open source http

### PR DESCRIPTION
**Pull request by jwoudenberg-nri**
One thing I want to do before calling this done: Remove `IORef` use from mocking API.

Most of this is git-subtreed out of our other repo, but there's a couple of cleanup and documentation commits towards the end.

### CODE REVIEW

File: nri-http/src/Internal/Http.hs [hunk 1]
```diff
@@ -42,14 +42,19 @@ data Expect a where
...
+-- - 'BadUrl' means you did not provide a valid URL.
+-- - 'Timeout' means it took too long to get a response.
+-- - 'NetworkError' means the user turned off their wifi, went in a cave, etc.
+-- - 'BadStatus' means you got a response back, but the status code indicates failure.
+-- - 'BadBody' means you got a response back with a nice status code, but the body of the response was something unexpected. The 'Text' in this cse is the debugging message that explains what went wrong with your JSONT decoder or whatever.
 data Error
   = BadUrl Text
-  | BadStatus Int
-  | BadBody Text
-  | BadResponse Text
```
Comment (stoeffel) [inline nri-http/src/Internal/Http.hs]: Is there a reason for removing `BadResponse`?

File: nri-http/src/Http.hs [hunk 2]
```diff
@@ -189,7 +189,12 @@ _request doAnythingHandler manager settings = do
                     Just mimeType ->
                       ("content-type", mimeType) : Internal.Http.headers settings,
                   HTTP.requestBody = HTTP.RequestBodyLBS <| Internal.Http.bodyContents (Internal.Http.body settings),
-                  HTTP.responseTimeout = HTTP.responseTimeoutMicro <| fromIntegral <| Maybe.withDefault (30 * 1000 * 1000) (Internal.Http.timeout settings)
+                  HTTP.responseTimeout =
+                    Internal.Http.timeout settings
+                      |> Maybe.withDefault (30 * 1000)
+                      |> (*) 1000
```
Comment (stoeffel) [inline nri-http/src/Http.hs:205]: Was this a 🐛 before?

**Review by stoeffel — COMMENTED** [review]
Looks great! I left a few comments though

File: nri-http/src/Http.hs (hunk 2 above)
Comment (jwoudenberg-nri) [reply nri-http/src/Http.hs:205]: Not really. We didn't document the timeout property before. This changes it from microseconds to milliseconds, to make it consistent with durations specified in other elm-ish functions (which typically take a `Float` in milliseconds).

File: nri-http/src/Internal/Http.hs (hunk 1 above)
Comment (jwoudenberg-nri) [reply nri-http/src/Internal/Http.hs]: Two!
- We weren't creating a `NetworkError` anywhere.
- The `BadResponse` constructor isn't one that exists in `elm/http`. Removing it in favor of `NetworkError` means our `Error` type is now the same as the `elm/http` one.

**Pull request merged**